### PR TITLE
[#20]Instr/IndexOf support BinaryString as input parameter

### PR DESCRIPTION
**Pull request by Kwafoor**
### What changes were proposed in this pull request?
Implemented the indexOf function in BinaryString by referencing spark UTF8String. Instr/IndexOf support BinaryString as input parameter.

### How was this PR tested?
- [x] Tests have Added for the changes
- [ ] Production environment verified

### CODE REVIEW

**Comment by pengzhiwei2018** [comment]
@Leomrlin  can you take a look at this PR first?

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/Instr.java [hunk 1]
```diff
@@ -44,4 +45,29 @@ public Long eval(String str, String target, Long from, Long nth) {
...
+
+    public Long eval(BinaryString str, BinaryString target, Long from, Long nth) {
+        if (str == null || target == null || from == null || nth == null) {
+            return null;
+        }
+        if (nth <= 0) {
+            return null;
+        }
+        int fromIndex = from.intValue() - 1;
+        if (fromIndex < 0) {
```
Comment (Leomrlin) [inline geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/Instr.java:65]: Why do we need to return null here?

**Review by Leomrlin — APPROVED** [review]
It looks good! Thank you for adding relevant comments and a lot of test cases. However, there is a small issue that may require your feedback.

File: geaflow/geaflow-common/src/main/java/com/antgroup/geaflow/common/binary/BinaryString.java [hunk 2]
```diff
@@ -226,6 +226,33 @@ public boolean contains(final BinaryString substring) {
         return false;
     }
 
+    public int indexOf(BinaryString v, int start) {
```
Comment (pengzhiwei2018) [inline geaflow/geaflow-common/src/main/java/com/antgroup/geaflow/common/binary/BinaryString.java]: Maybe the variable "s" or "str" is more reasonable than "v"

**Review by pengzhiwei2018 — COMMENTED** [review]
@Kwafoor Thanks for you contribution. Nice work! After some minor comments fixed, we can merge it.

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/Instr.java (hunk 1 above)
Comment (beryllw) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/Instr.java:65]: > Why do we need to return null here?

How about returning -1 to be consistent with `IndexOf`? Returning null is consistent with the previous code `Instr`.

File: geaflow/geaflow-common/src/main/java/com/antgroup/geaflow/common/binary/BinaryString.java (hunk 2 above)
Comment (beryllw) [reply geaflow/geaflow-common/src/main/java/com/antgroup/geaflow/common/binary/BinaryString.java]: Ok, I'll make the change.

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/Instr.java (hunk 1 above)
Comment (Leomrlin) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/Instr.java:65]: There may be multiple interpretations here. My understanding is that fromIndex < 0 can be regarded as starting the search from the beginning of the string. However, in Aliyun's ODPS, a negative start_position is considered to indicate that the starting position is counted backwards from the end of the string. Refer to https://help.aliyun.com/document_detail/455546.html?spm=a2c4g.48973.0.i1 for more information.

Comment (pengzhiwei2018) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/Instr.java:65]: @Kwafoor @Leomrlin The function **Instr** is defined the Hive SQL and most SQL Engine on BigData adopt it. So we should also follow the semantics of hive(ODPS also follow this semantics).   I will do  this in another PR. So we can stop the arguing for `fromIndex < 0` here.

Comment (Leomrlin) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/Instr.java:65]: ok, that's great. It's not a problem.

**Review by pengzhiwei2018 — APPROVED** [review]
LGTM

**Pull request merged**